### PR TITLE
Switch pylibcudf-enabled types to use enum class in Cython

### PR DESCRIPTION
**Pull request by vyasr**
## Description

This PR leverages Cython 3's support for C++'s scoped enumerations to more faithfully translate libcudf types into cuDF Cython and pylibcudf. Due to some Cython 3 limitations, there are a few workarounds in this PR. However, they are relatively minor and can be adjusted later as Cython improves. In the meantime, the change here is an improvement still worth merging, especially since it sets a template for how pylibcudf code should look as more of it emerges. Existing cuDF Cython is only updated to the minimum extent required for it to compile. Fully switching the old code to use enum class-style syntax isn't worthwhile since those internals should eventually be migrated to use pylibcudf in pure Python mode anyway.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by shwina — APPROVED** [review]
LGTM. Hopefully we have fixes in Cython in soon that will ease some of the pain here.

File: docs/cudf/source/developer_guide/pylibcudf.md [hunk 1]
```diff
@@ -86,26 +86,70 @@ cpdef Table gather(
...
+Our usage of enums is intended to minimize the complexity of our code while also working around Cython's limitations.
+
+```{warning}
+The guidance in this section may change often as Cython is updated and our understanding of best practices evolves.
+```
+
+- All pxd files that declare a C++ enum should use `cpdef enum class` declarations.
+  -  Reason: This declaration makes the C++ enum available in Cython code while also transparently creating a Python enum.
+- Any pxd file containing only C++ declarations must still have a corresponding pyx file if any of the declarations are scoped enums.
+  - Reason: The creation of the Python enum requires that Cython actually generate the necessary code Python C API code, which will not happen if only a pxd file is present.
```
Comment (bdice) [inline docs/cudf/source/developer_guide/pylibcudf.md:113]:
```suggestion
  - Reason: The creation of the Python enum requires that Cython actually generate the necessary Python C API code, which will not happen if only a pxd file is present.
```

**Review by bdice — APPROVED** [review]
Looks fine to me. I hope we are able to shed the tech debt (issue workarounds) sooner than later...

**Pull request merged**